RevokeCgn/handler.ts

```diff
@@ -0,0 +1,170 @@
+import {
+  CgnRevokedStatus,
+  StatusEnum
+} from "../generated/definitions/CgnRevokedStatus";
+import { InstanceId } from "../generated/definitions/InstanceId";
+import { RetrievedUserCgn, UserCgnModel } from "../models/user_cgn";
+import { OrchestratorInput } from "../UpdateCgnOrchestrator";
+import { makeUpdateCgnOrchestratorId } from "../utils/orchestrators";
+import { checkUpdateCgnIsRunning } from "../utils/orchestrators";
+
+type ErrorTypes =
+  | IResponseErrorInternal
+  | IResponseErrorNotFound
+  | IResponseErrorConflict;
+type ReturnTypes =
+  | IResponseSuccessAccepted
+  | IResponseSuccessRedirectToResource<InstanceId, InstanceId>
+  | ErrorTypes;
+
+type IRevokeCgnHandler = (
+  context: Context,
+  fiscalCode: FiscalCode,
+  cgnRevokationRequest: CgnRevokationRequest
+) => Promise<ReturnTypes>;
+
+const checkExistingCgnStatus = (userCgn: RetrievedUserCgn) =>
+  CgnRevokedStatus.is(userCgn.status) || CgnCanceledStatus.is(userCgn.status)
+    ? fromLeft<IResponseErrorConflict, RetrievedUserCgn>(
+        ResponseErrorConflict(
+          "Cannot revoke the user's cgn because it is already revoked or canceled"
+        )
+      )
+    : taskEither.of<IResponseErrorConflict, RetrievedUserCgn>(userCgn);
+
+export function RevokeCgnHandler(
+  userCgnModel: UserCgnModel,
+  logPrefix: string = "RevokeCgnHandler"
+): IRevokeCgnHandler {
+  return async (context, fiscalCode, revokationReq) => {
+    const client = df.getClient(context);
+    const revokedCgnStatus: CgnRevokedStatus = {
+      motivation: revokationReq.motivation,
+      revokation_date: new Date(),
+      status: StatusEnum.REVOKED
+    };
+    const orchestratorId = makeUpdateCgnOrchestratorId(
+      fiscalCode,
+      StatusEnum.REVOKED
+    ) as NonEmptyString;
+    return userCgnModel
+      .findLastVersionByModelId([fiscalCode])
+      .mapLeft<IResponseErrorInternal | IResponseErrorNotFound>(() =>
+        ResponseErrorInternal("Cannot retrieve CGN infos for this user")
+      )
+      .chain(maybeUserCgn =>
+        fromEither(
+          fromOption(
+            ResponseErrorNotFound("Not Found", "User's CGN status not found")
+          )(maybeUserCgn)
+        )
+      )
+      .foldTaskEither<ErrorTypes, RetrievedUserCgn>(fromLeft, userCgn =>
+        checkExistingCgnStatus(userCgn)
+      )
+      .map(userCgn => ({
+        ...userCgn,
+        status: revokedCgnStatus
+      }))
+      .chain(_ =>
+        userCgnModel
+          .update(_)
+          .mapLeft(err =>
+            ResponseErrorInternal(
+              `Could not update Cgn model ${toError(err).message}`
+            )
+          )
+      )
+      .chain(() =>
+        checkUpdateCgnIsRunning(
+          client,
+          fiscalCode,
+          revokedCgnStatus
+        ).foldTaskEither<
+          ErrorTypes,
+          | IResponseSuccessAccepted
+          | IResponseSuccessRedirectToResource<InstanceId, InstanceId>
+        >(
```

Comment:
Based on this approach, we perform model update into Http handler (as suggested by @balanza ) and let the orchestrator do remaining job steps (in this case only Send Message). I'm wondering about `checkUpdateCgnIsRunning` ErrorTypes because, if we assume that the update operation works fine, than we can either return a 500 Error due to errors while starting orchestrators. In this case the caller while trying to re-invoke the same API will obtain 409 ( as the CGN is already in a `REVOKED` ). On the other side if we consider orchestrator's execution as a "Fire&Forget" operation, we don't take care about possible errors in durable functions client start. What do you think about this? @gunzip @balanza @BurnedMarshal